contracts/OtokenFactory.sol

```diff
@@ -0,0 +1,185 @@
+pragma solidity =0.6.10;
+
+import {OtokenSpawner} from "./OtokenSpawner.sol";
+import {AddressBookInterface} from "./interfaces/AddressBookInterface.sol";
+import {OtokenInterface} from "./interfaces/OtokenInterface.sol";
+import {WhitelistInterface} from "./interfaces/WhitelistInterface.sol";
+
+/**
+ * SPDX-License-Identifier: UNLICENSED
+ * @title A factory for opyn tokens
+ * @author Opyn
+ * @notice Create new otokens and keep track of all created tokens.
+ * @dev Calculate contract address before each creation with CREATE2
+ * and deploy eip-1167 minimal proxies for otoken logic contract.
+ */
+contract OtokenFactory is OtokenSpawner {
+    /// @notice The Opyn AddressBook contract that records addresses of whitelist module and otoken impl address. */
+    address public addressBook;
+
+    /// @dev An array of all created otokens */
+    address[] private _otokens;
+
+    /// @dev A mapping from parameters hash to its deployed address
+    mapping(bytes32 => address) private _idToAddress;
+
+    constructor(address _addressBook) public {
+        addressBook = _addressBook;
+    }
+
+    /* Event: OtokenCreated - Emitted when factory create a new Option. */
+    event OtokenCreated(
+        address indexed tokenAddress,
+        address indexed creator,
+        address underlying,
+        address strike,
+        address collateral,
+        uint256 strikePrice,
+        uint256 expiry,
+        bool isPut
+    );
+
+    /**
+     * @notice create new otokens
+     * @dev deploy an eip-1167 minimal proxy with CREATE2 and register it to the whitelist module.
+     * @param _underlyingAsset asset that the option references
+     * @param _strikeAsset asset that the strike price is denominated in
+     * @param _collateralAsset asset that is held as collateral against short/written options
+     * @param _strikePrice strike price with decimals = 18
+     * @param _expiry expiration timestamp in second
+     * @param _isPut is this a put option, if not it is a call
+     * @return newOtoken address of the newly created option
+     */
+    function createOtoken(
+        address _underlyingAsset,
+        address _strikeAsset,
+        address _collateralAsset,
+        uint256 _strikePrice,
+        uint256 _expiry,
+        bool _isPut
+    ) external returns (address newOtoken) {
+        require(_expiry > now, "OtokenFactory: Can't create expired option.");
+        bytes32 id = _getOptionId(_underlyingAsset, _strikeAsset, _collateralAsset, _strikePrice, _expiry, _isPut);
+        require(_idToAddress[id] == address(0), "OtokenFactory: Option created");
```

Comment:
for these two, can you call the getOToken function and make that a public function? When should we make functions external vs public? @haythem96

Reply:
Yeah those two lines can be refactored by calling the `getOtoken`.
In this case L62 and `_idToAddress[id] ` are replicating the getOtoken function, so it make sense to make it public and call it.

Reply:
I use the `id` here because in the later part of the function, I need to set
```
_idToAddress[id] = newOtoken;
```

I don't want to calculate the id twice.

Reply:
mm yeah make sense (y).